src/webui/www/private/search.html

```diff
@@ -0,0 +1,632 @@
+                    loadSearchPlugins();
+                },
+                onClose: function() {
+                    clearTimeout(loadSearchPluginsTimer);
+                }
+            });
+    };
+
+    loadSearchPlugins = function() {
+        getPlugins();
+        loadSearchPluginsTimer = loadSearchPlugins.delay(2000);
+    };
+
+    categorySelected = function() {
+        selectedCategory = $("categorySelect").get("value");
+    };
+
+    pluginSelected = function() {
+        selectedPlugin = $("pluginsSelect").get("value");
+
+        if (selectedPlugin !== prevSelectedPlugin) {
+            prevSelectedPlugin = selectedPlugin;
+            getSearchCategories();
+        }
+    };
+
+    reselectCategory = function() {
+        for (var i = 0; i < $("categorySelect").options.length; i++)
+            if ($("categorySelect").options[i].get("value") === selectedCategory)
+                $("categorySelect").options[i].selected = true;
+
+        categorySelected();
+    };
+
+    reselectPlugin = function() {
+        for (var i = 0; i < $("pluginsSelect").options.length; i++)
+            if ($("pluginsSelect").options[i].get("value") === selectedPlugin)
+                $("pluginsSelect").options[i].selected = true;
+
+        pluginSelected();
+    };
+
+    resetSearchState = function() {
+        clearTimeout(loadSearchResultsTimer);
+        $('startSearchButton').set('text', 'QBT_TR(Search)QBT_TR[CONTEXT=SearchEngineWidget]');
+        searchResultsRowId = 0;
+        searchRunning = false;
+        activeSearchId = null;
+    };
+
+    getSearchCategories = function() {
+        var populateCategorySelect = function(categories) {
+            var categoryHtml = [];
+            categories.each(function(category) {
+                categoryHtml.push("<option>" + category + "</option>");
+            });
+
+            // first category is "All Categories"
+            if (categoryHtml.length > 1)
+                // add separator
+                categoryHtml.splice(1, 0, "<option disabled>──────────</option>");
+
+            $('categorySelect').set('html', categoryHtml.join(""));
+        };
+
+        var selectedPlugin = $('pluginsSelect').get("value");
+        if ((selectedPlugin === "all") || (selectedPlugin === "enabled")) {
+            var url = new URI('api/v2/search/categories');
+            url.setData('name', selectedPlugin);
+            var request = new Request.JSON({
+                url: url,
+                noCache: true,
+                method: 'get',
+                onSuccess: function(response) {
+                    populateCategorySelect(response);
+                }
+            }).send();
+        }
+        else {
+            var plugin = getPlugin(selectedPlugin);
+            var plugins = [];
+
+            plugins.push("QBT_TR(All categories)QBT_TR[CONTEXT=SearchEngineWidget]");
+            if (plugin !== null)
+                plugins = plugins.concat(plugin.supportedCategories);
```

Comment:
why not just `plugins.push();`?

Reply:
`plugin.supportedCategories` is an array, so `push()` will add the entire array as an element of `plugins`. `concat()` will instead individually all elements of the array to `plugins`.